presto-iceberg/src/main/java/com/facebook/presto/iceberg/optimizer/IcebergMetadataOptimizer.java

```diff
@@ -0,0 +1,411 @@
+            return new ProjectNode(node.getSourceLocation(), idAllocator.getNextId(), valuesNode, assignments, LOCAL);
+        }
+
+        private RowExpression evaluateMinMax(FunctionMetadata aggregationFunctionMetadata, List<RowExpression> arguments)
+        {
+            Type returnType = typeManager.getType(aggregationFunctionMetadata.getReturnType());
+            if (arguments.isEmpty()) {
+                return new ConstantExpression(Optional.empty(), null, returnType);
+            }
+
+            String scalarFunctionName = AGGREGATION_SCALAR_MAPPING.get(aggregationFunctionMetadata.getName()).getObjectName();
+            while (arguments.size() > 1) {
+                List<RowExpression> reducedArguments = new ArrayList<>();
+                // We fold for every 100 values because GREATEST/LEAST has argument count limit
+                for (List<RowExpression> partitionedArguments : Lists.partition(arguments, 100)) {
+                    FunctionHandle functionHandle;
+                    if (scalarFunctionName.equals("greatest")) {
+                        functionHandle = functionResolution.greatestFunction(partitionedArguments.stream().map(RowExpression::getType).collect(toImmutableList()));
+                    }
+                    else if (scalarFunctionName.equals("least")) {
+                        functionHandle = functionResolution.leastFunction(partitionedArguments.stream().map(RowExpression::getType).collect(toImmutableList()));
+                    }
+                    else {
+                        throw new PrestoException(StandardErrorCode.NOT_SUPPORTED, "unsupported function: " + scalarFunctionName);
+                    }
+
+                    Object reducedValue = rowExpressionService.getExpressionInterpreter().evaluateConstantRowExpression(
+                            new CallExpression(
+                                    Optional.empty(),
+                                    scalarFunctionName,
+                                    functionHandle,
+                                    returnType,
+                                    arguments),
+                            connectorSession);
+                    reducedArguments.add(new ConstantExpression(reducedValue, returnType));
+                }
+                arguments = reducedArguments;
+            }
+            return getOnlyElement(arguments);
+        }
+
+        private static Optional<TableScanNode> findTableScan(PlanNode source, DeterminismEvaluator determinismEvaluator)
+        {
+            while (true) {
+                // allow any chain of linear transformations
+                if (source instanceof MarkDistinctNode ||
+                        source instanceof FilterNode ||
+                        source instanceof SortNode) {
```

Comment:
`Limit` could not fit here, referring to PR #22203 . As for `DistinctLimit`, seems it should be in the same position as `Aggregation`, maybe we should implement the method `visitDistinctLimit` to support it. I'm not so sure, would figure it out.

Reply:
Seems `DistinctLimitNode` is in the same position as `AggregationNode`. So we should implement method `visitDistinctLimit` to support it. I need more time to confirm its fitness for this scenario, as its already involving in several optimizations. So maybe a separate PR to implement this would be better, what do your think?

Reply:
Sure, we can just leave to a later PR. I had gone through the list of nodes and I wanted to check that we weren't leaving any out that could have been added to this conditional